### ISSUE TITLE
Probable memory corruption when running the unit tests (under MacOSX)

### ISSUE THREAD

**Issue by SkyToGround** [ess-dmsc/h5cpp#108]
I am seeing frequent *EXC_BAD_ACCESS* and pointer errors when running the unit tests under MacOSX, but only when compiling without debug symbols. Some of the errors are shown below. Due to the nature of the errors I believe that they are caused by bugs in the code that for now have only manifested under MacOSX but that are also present under Linux.

**Run 1**
```
15 FAILED TESTS
Process 14348 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00000001003f1ed0 libhdf5.101.dylib`H5FL_garbage_coll + 320
libhdf5.101.dylib`H5FL_garbage_coll:
->  0x1003f1ed0 <+320>: movq   0x8(%rsi), %rdi
    0x1003f1ed4 <+324>: movq   0x10(%rsi), %r15
    0x1003f1ed8 <+328>: testq  %rdi, %rdi
    0x1003f1edb <+331>: je     0x1003f1f2d               ; <+413>
Target 0: (unit_tests) stopped.
```

**Run 2**
```
Process 14384 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x464646)
    frame #0: 0x00000001003f1fc0 libhdf5.101.dylib`H5FL_garbage_coll + 560
libhdf5.101.dylib`H5FL_garbage_coll:
->  0x1003f1fc0 <+560>: movq   (%rdi), %rbx
    0x1003f1fc3 <+563>: decl   0x4(%r12)
    0x1003f1fc8 <+568>: callq  0x1005cd698               ; symbol stub for: free
    0x1003f1fcd <+573>: testq  %rbx, %rbx
Target 0: (unit_tests) stopped.
```

**Run 3**
```
unit_tests(14423,0x7fff912fe340) malloc: *** error for object 0x101464646: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Process 14423 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff57d83fce libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff57d83fce <+10>: jae    0x7fff57d83fd8            ; <+20>
    0x7fff57d83fd0 <+12>: movq   %rax, %rdi
    0x7fff57d83fd3 <+15>: jmp    0x7fff57d7b76c            ; cerror_nocancel
    0x7fff57d83fd8 <+20>: retq
Target 0: (unit_tests) stopped.
```

**Run 4**
```
unit_tests(14442,0x7fff912fe340) malloc: *** error for object 0x10075e280: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
Process 14442 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff57d83fce libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff57d83fce <+10>: jae    0x7fff57d83fd8            ; <+20>
    0x7fff57d83fd0 <+12>: movq   %rax, %rdi
    0x7fff57d83fd3 <+15>: jmp    0x7fff57d7b76c            ; cerror_nocancel
    0x7fff57d83fd8 <+20>: retq
Target 0: (unit_tests) stopped.
```

**Comment by eugenwintersberger** [comment]
My first guess would be that this might be related to a problem I currently have on Linux and Windows on branch 100. Currently I get a segmentation fault on Linux and a SEH exception on Windows. There is something bad lurking in the darkness of the code ;) And we definitely have to face this evil bastard ;)

**Comment by SkyToGround** [comment]
This appears to be fixed on #109. I will leave this issue open until that code has been merged into the master branch however.
